src/app/App.js

```diff
@@ -1310,7 +1322,7 @@ class App extends EventEmitter {
 			}
 		};
 
-		return new CancellablePromise((resolve) => sync() & async.nextTick(() => sync() & resolve()));
+		return new Promise((resolve) => sync() & async.nextTick(() => sync() & resolve()));
```

Comment:
Mentioned this to you offline because this line is very non-obvious. You said you have to loop back and deal with `async` here.

Reply:
Yes, so I looked into this and I'm a bit confused mostly because of how [`async`](https://github.com/metal/metal.js/blob/17e5fb69956a5c14d367addc1af0898a89cff42e/packages/metal/src/async/async.js) is written: it uses `MessageChannel` (which is something I've never used) and then falls back to whatever is available, in somecases that just means `setTimeout(fn, 0)`. I'm surprised we've been using this for so long given the fact that we didn't even write it but  `¯\_(ツ)_/¯`.

I re-wrote this bit: 

```javacript
return new Promise((resolve) => sync() & async.nextTick(() => sync() & resolve()));
```

as this:

```javacript
return new Promise((resolve) => {
    return sync() & (setTimeout(() => {
        return sync() & resolve();
    }), 0);
});
```

And it _looks_ like it's working in the same way, but I'm not 100% sure, so just for the sake of safety I don't want to remove `async` just yet, although I definetly think that we should look for an alternative (maintaining or making changes to this project is not an easy task).

Reply:
I don't think it makes sense to use the [bitwise AND operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators#Bitwise_AND) (`&`) here.  That to me looks like a mistake in the original version of the code (perhaps the author meant to write `&&`); either that or it is an obfuscated attempt at doing something like what [the comma operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator) does.

In any case, `setTimeout` returns a timer ID, `resolve()` returns `undefined`, and it doesn't seem to make any sense to `&` those with whatever `sync()` returns (I don't know what that is). Likewise, returning a value from a `setTimeout()` callback makes no sense, because there is no "caller" to evaluate the returned value.

In other words, I suspect this could just be:

```javascript
return new Promise(resolve => {
  sync();
  setTimeout(() => {
    sync();
    resolve();
  }, 0);
});
```

Which seems to match the intent expressed in the commit (c2f5b618e5f555bc2eab79e99c7846dea97a105f) that added the original line.

Reply:
(Edited to fix commit link.)

Reply:
Thanks for the details @wincent, that made it clearer.